koku/masu/util/common.py

```diff
@@ -455,3 +457,43 @@ def __call__(cls, *args, **kwargs):
                 instance = super().__call__(*args, **kwargs)
                 cls._instances[cls] = instance
         return cls._instances[cls]
+
+
+def check_setup_complete(provider_uuid):
+    with ProviderDBAccessor(provider_uuid=provider_uuid) as provider_accessor:
+        return provider_accessor.get_setup_complete()
+
+
+def get_provider_updated_timestamp(provider_uuid):
+    with ProviderDBAccessor(provider_uuid=provider_uuid) as provider_accessor:
+        return provider_accessor.get_data_updated_timestamp()
+
+
+def get_start_delta(start_date, provider_uuid):
+    """Get initial start date for download file processing"""
+    dh = DateHelper()
+    if start_date.year == dh.today.year and start_date.month == dh.today.month:
+        last_update_day = get_provider_updated_timestamp(provider_uuid)
+        if not last_update_day:
+            last_update_day = dh.today
+        start_delta = (
+            last_update_day - datetime.timedelta(days=3) if last_update_day.day > 3 else last_update_day.replace(day=1)
+        )
+    else:
+        start_delta = dh.month_end(start_date) - datetime.timedelta(days=3)
+    start_delta = start_delta.replace(tzinfo=None)
+    return start_delta
+
+
+def fetch_optional_columns(local_file, current_columns, fetch_columns, tracing_id, context):
+    """Add optional columns to columns list if they exists in files"""
+    for fetch_column in fetch_columns:
+        try:
+            data_frame = pd.read_csv(local_file, usecols=lambda col: col.lower().startswith(fetch_column))
```

Comment:
How large is this CSV file? Should it be read in chunks in order to avoid hitting memory limits?

Reply:
Is there a way to only read the CSV file once rather than on each iteration through the loop?

Reply:
:confused: So that is part of the point of this to save memory overall by not loading the whole file, we are loading a all columns that start with 'tags'  for example. For one customer we have 1000 tag columns and 950 have nulls in all rows. This grabs specific columns and drops any that have dumb data. :+1: